source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -84,6 +84,10 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
     int rc = SSL_CTX_set_app_data(ctx.ssl_ctx_.get(), this);
     RELEASE_ASSERT(rc == 1, Utility::getLastCryptoError().value_or(""));
 
+    constexpr uint32_t mode = SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER;
+    rc = SSL_CTX_set_mode(ctx.ssl_ctx_.get(), mode);
```

Comment:
When is this needed? `maybeLinearize` returns either unmodified slice or linearized slice, but I don't think there is a case when unmodified slice would be later linearized (assuming `desired_min_size <= max_size`), so the buffer shouldn't move... or am I missing something?

Reply:
It probably isn't strictly necessary, and with the current implementation I don't think the buffer will move. But if the `maybeLinearize` interface doesn't guarantee this property, and I didn't want to keep the existing book-keeping to ensure the write buffer doesn't change. It doesn't look to me like anything is more expensive in boringSSL when this mode is set, it just removes a check that isn't gaining anything for how we use the API.

Reply:
Sure, but if `maybeLinearize` implementation changes enough to require moving buffers, then we can enable this mode. Right now, it removes the default sanity check for no reason.

Reply:
Huh, looking into this I realize I lost part of the change (I did a bunch of moving code between branches to chop an originally big change into manageable pieces). To simplify this code, I removed `bytes_to_retry_`. I subsequent call can then end up with a larger buffer from `maybeLinearize` if data was added to the buffer since the last attempt at `SSL_write`. That's why I was setting this option here. 

I'm trying to remember why I made that change originally; I think it may have been to make it easier to read and reason about. I don't think it had a measurable performance impact. Any preference on whether I make that change or not?

Reply:
> I subsequent call can then end up with a larger buffer from maybeLinearize if data was added to the buffer since the last attempt at SSL_write. That's why I was setting this option here.

I don't believe that `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` allows for buffer to grow between retries. AFAIK, the buffer data has to stay the same, but it can be available at a different address than before.

Reply:
The docs say it is allowed: 
```
// In TLS, a non-blocking |SSL_write| differs from non-blocking |write| in that
// a failed |SSL_write| still commits to the data passed in. When retrying, the
// caller must supply the original write buffer (or a larger one containing the
// original as a prefix). By default, retries will fail if they also do not
// reuse the same |buf| pointer. This may be relaxed with
// |SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER|, but the buffer contents still must be
// unchanged.
```

Reply:
The last sentence literally says `the buffer contents still must be unchanged.`

Reply:
...but there is also `(or a larger one containing the original as a prefix)`, hmm. Maybe it's allowed after all.